allennlp/models/vision_text_model.py

```diff
@@ -0,0 +1,256 @@
+from allennlp.modules.transformer import (
+    TextEmbeddings,
+    ImageFeatureEmbeddings,
+    BiModalEncoder,
+    TransformerPooler,
+)
+
+from transformers.modeling_auto import AutoModel
+
+logger = logging.getLogger(__name__)
+
+
+@Model.register("vision_model")
+class VisionTextModel(Model):
+    """
+    `VisionTextModel` takes as input a single text input and a single image input
+    to produce some output. Example tasks include visual question-answering, visual
+    entailment, etc.
+
+    # Parameters
+
+    vocab : `Vocabulary`
+    text_embeddings : `TextEmbeddings`
+    image_embeddings : `ImageFeatureEmbeddings`
+    encoder : `BiModalEncoder`
+    pooled_output_dim : `int`
+    fusion_method : `str`, optional (default = `"sum"`)
+    dropout : `float`, optional (default = `0.1`)
+    label_namespace : `str`, optional (default = `"labels"`)
+    """
+
+    def __init__(
+        self,
+        vocab: Vocabulary,
+        text_embeddings: TextEmbeddings,
+        image_embeddings: ImageFeatureEmbeddings,
+        encoder: BiModalEncoder,
+        pooled_output_dim: int,
+        fusion_method: str = "sum",
+        dropout: float = 0.1,
+        label_namespace: str = "labels",
+    ) -> None:
+
+        super().__init__(vocab)
+
+        self.fusion_method = fusion_method
+
+        self.embeddings = text_embeddings
+        self.image_embeddings = image_embeddings
+        self.encoder = encoder
+
+        self.t_pooler = TransformerPooler(encoder.hidden_size1, pooled_output_dim)
+        self.v_pooler = TransformerPooler(encoder.hidden_size2, pooled_output_dim)
+
+        num_labels = vocab.get_vocab_size(label_namespace)
+        self.label_namespace = label_namespace
+
+        self.classifier = torch.nn.Linear(pooled_output_dim, num_labels)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    @classmethod
+    def from_huggingface_model_name(
+        cls,
+        vocab: Vocabulary,
+        model_name: str,
+        image_feature_dim: int,
+        image_num_hidden_layers: int,
+        image_hidden_size: int,
+        image_num_attention_heads: int,
+        combined_hidden_size: int,
+        combined_num_attention_heads: int,
+        pooled_output_dim: int,
+        image_intermediate_size: int,
+        image_attention_dropout: float,
+        image_hidden_dropout: float,
+        image_biattention_id: List[int],
+        text_biattention_id: List[int],
+        text_fixed_layer: int,
+        image_fixed_layer: int,
+        pooled_dropout: float = 0.1,
+        fusion_method: str = "sum",
+    ):
+        transformer = AutoModel.from_pretrained(model_name)
+
+        text_embeddings = deepcopy(transformer.embeddings)
+
+        # Albert (and maybe others?) has this "embedding_size", that's different from "hidden_size".
+        # To get them to the same dimensionality, it uses a linear transform after the embedding
+        # layer, which we need to pull out and copy here.
```

Comment:
We still need to do this, right? This is not abstracted away by the Transformer Toolkit?

Reply:
Not yet. Hopefully soon.